cFrom: csagan5 <32685696+csagan5@users.noreply.github.com>
Date: Mon, 18 Mar 2019 21:47:12 +0100
Subject: Add option to use home page as NTP

Use about:blank as default homepage
---
 chrome/android/java/res/xml/homepage_preferences.xml                                           |    5 ++
 chrome/android/java/src/org/chromium/chrome/browser/partnercustomizations/HomepageManager.java |   24 +++++++++-
 chrome/android/java/src/org/chromium/chrome/browser/settings/HomepagePreferences.java          |   10 ++++
 chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java            |   11 +++-
 chrome/browser/ui/android/strings/android_chrome_strings.grd                                   |    3 +
 chrome/browser/ui/browser_ui_prefs.cc                                                          |    2 
 chrome/common/pref_names.cc                                                                    |    4 +
 chrome/common/pref_names.h                                                                     |    1 
 8 files changed, 56 insertions(+), 4 deletions(-)

--- a/chrome/android/java/res/xml/homepage_preferences.xml
+++ b/chrome/android/java/res/xml/homepage_preferences.xml
@@ -7,6 +7,11 @@
     xmlns:android="http://schemas.android.com/apk/res/android">
 
     <org.chromium.chrome.browser.settings.ChromeSwitchPreference
+        android:key="ntp_is_homepage_switch"
+        android:summaryOn="@string/options_ntp_is_homepage_label"
+        android:summaryOff="@string/options_ntp_is_homepage_label" />
+
+    <org.chromium.chrome.browser.settings.ChromeSwitchPreference
         android:key="homepage_switch"
         android:summaryOn="@string/text_on"
         android:summaryOff="@string/text_off" />
--- a/chrome/android/java/src/org/chromium/chrome/browser/partnercustomizations/HomepageManager.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/partnercustomizations/HomepageManager.java
@@ -31,6 +31,7 @@ public class HomepageManager {
     }
 
     private static final String PREF_HOMEPAGE_ENABLED = "homepage";
+    public static final String PREF_NTP_IS_HOMEPAGE = "newtabpage_is_homepage";
     private static final String PREF_HOMEPAGE_CUSTOM_URI = "homepage_custom_uri";
     private static final String PREF_HOMEPAGE_USE_DEFAULT_URI = "homepage_partner_enabled";
 
@@ -90,7 +91,8 @@ public class HomepageManager {
      */
     public static boolean shouldCloseAppWithZeroTabs() {
         return HomepageManager.isHomepageEnabled()
-                && !NewTabPage.isNTPUrl(HomepageManager.getHomepageUri());
+                && !NewTabPage.isNTPUrl(HomepageManager.getHomepageUri())
+                && (HomepageManager.getHomepageUri() != "about:blank");
     }
 
     /**
@@ -113,7 +115,7 @@ public class HomepageManager {
     public static String getDefaultHomepageUri() {
         return PartnerBrowserCustomizations.isHomepageProviderAvailableAndEnabled()
                 ? PartnerBrowserCustomizations.getHomePageUrl()
-                : UrlConstants.NTP_NON_NATIVE_URL;
+                : "about:blank";
     }
 
     /**
@@ -127,6 +129,14 @@ public class HomepageManager {
     }
 
     /**
+     * Returns the user preference for whether the New Tab Page is the homepage or not.
+     *
+     */
+    public boolean getPrefNTPIsHomepageEnabled() {
+        return mSharedPreferences.getBoolean(PREF_NTP_IS_HOMEPAGE, false);
+    }
+
+    /**
      * Sets the user preference for whether the homepage is enabled.
      */
     public void setPrefHomepageEnabled(boolean enabled) {
@@ -139,6 +149,16 @@ public class HomepageManager {
         notifyHomepageUpdated();
     }
 
+    /**
+     * Sets the user preference for whether the new tab page is the homepage or not.
+     */
+    public void setPrefNTPIsHomepageEnabled(boolean enabled) {
+        SharedPreferences.Editor sharedPreferencesEditor = mSharedPreferences.edit();
+        sharedPreferencesEditor.putBoolean(PREF_NTP_IS_HOMEPAGE, enabled);
+        sharedPreferencesEditor.apply();
+        notifyHomepageUpdated();
+    }
+
     /**
      * @return User specified homepage custom URI string.
      */
--- a/chrome/android/java/src/org/chromium/chrome/browser/settings/HomepagePreferences.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/settings/HomepagePreferences.java
@@ -21,6 +21,7 @@ public class HomepagePreferences extends
     @VisibleForTesting
     public static final String PREF_HOMEPAGE_SWITCH = "homepage_switch";
     private static final String PREF_HOMEPAGE_EDIT = "homepage_edit";
+    private static final String PREF_NTP_HOMEPAGE_SWITCH = "ntp_is_homepage_switch";
 
     private HomepageManager mHomepageManager;
     private Preference mHomepageEdit;
@@ -46,6 +47,15 @@ public class HomepagePreferences extends
             });
         }
 
+        ChromeSwitchPreference mNTPIsHomepageSwitch =
+                (ChromeSwitchPreference) findPreference(PREF_NTP_HOMEPAGE_SWITCH);
+        boolean isHomepageNTPEnabled = mHomepageManager.getPrefNTPIsHomepageEnabled();
+        mNTPIsHomepageSwitch.setChecked(isHomepageNTPEnabled);
+        mNTPIsHomepageSwitch.setOnPreferenceChangeListener((preference, newValue) -> {
+            mHomepageManager.setPrefNTPIsHomepageEnabled((boolean) newValue);
+            return true;
+        });
+
         mHomepageEdit = findPreference(PREF_HOMEPAGE_EDIT);
         updateCurrentHomepageUrl();
     }
--- a/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java
@@ -12,6 +12,7 @@ import org.chromium.chrome.browser.tab.T
 import org.chromium.chrome.browser.util.UrlConstants;
 import org.chromium.content_public.browser.LoadUrlParams;
 import org.chromium.content_public.browser.WebContents;
+import org.chromium.chrome.browser.partnercustomizations.HomepageManager;
 
 /**
  * An interface to return a {@link TabCreator} either for regular or incognito tabs.
@@ -85,12 +86,18 @@ public interface TabCreatorManager {
         }
 
         /**
-         * Creates a new tab and loads the NTP.
+         * Creates a new tab and loads the NTP or the homepage, depending on user preferences.
          */
         public final void launchNTP() {
             try {
+                String newTabURL;
+                if (HomepageManager.getInstance().getPrefNTPIsHomepageEnabled()) {
+                   newTabURL = HomepageManager.getHomepageUri();
+                } else {
+                   newTabURL = UrlConstants.NTP_URL;
+                }
                 TraceEvent.begin("TabCreator.launchNTP");
-                launchUrl(UrlConstants.NTP_URL, TabLaunchType.FROM_CHROME_UI);
+                launchUrl(newTabURL, TabLaunchType.FROM_CHROME_UI);
             } finally {
                 TraceEvent.end("TabCreator.launchNTP");
             }
--- a/chrome/browser/ui/android/strings/android_chrome_strings.grd
+++ b/chrome/browser/ui/android/strings/android_chrome_strings.grd
@@ -702,6 +702,9 @@ CHAR-LIMIT guidelines:
       <message name="IDS_OPTIONS_HOMEPAGE_EDIT_LABEL" desc="The label for the edit text field that allows the user to change the URL that is opened when they tap on the home page button in the omnibox.">
         Open this page
       </message>
+      <message name="IDS_OPTIONS_NTP_IS_HOMEPAGE_LABEL" desc="The label for switch that allows the user to toggle whether opening a new tab leads to the new tab page or the home page.">
+        Use for new tabs
+      </message>
 
       <!-- Notifications preferences -->
       <message name="IDS_PREFS_NOTIFICATIONS" desc="Title for Notification preferences.">
--- a/chrome/browser/ui/browser_ui_prefs.cc
+++ b/chrome/browser/ui/browser_ui_prefs.cc
@@ -55,6 +55,8 @@ void RegisterBrowserUserPrefs(user_prefs
                                 GetHomeButtonAndHomePageIsNewTabPageFlags());
   registry->RegisterBooleanPref(prefs::kShowHomeButton, false,
                                 GetHomeButtonAndHomePageIsNewTabPageFlags());
+  registry->RegisterBooleanPref(prefs::kNewTabPageIsHomePage, false,
+                                GetHomeButtonAndHomePageIsNewTabPageFlags());
 
   registry->RegisterInt64Pref(prefs::kDefaultBrowserLastDeclined, 0);
   bool reset_check_default = false;
--- a/chrome/common/pref_names.cc
+++ b/chrome/common/pref_names.cc
@@ -59,6 +59,10 @@ const char kForceEphemeralProfiles[] = "
 // A boolean specifying whether the New Tab page is the home page or not.
 const char kHomePageIsNewTabPage[] = "homepage_is_newtabpage";
 
+// A boolean specifying whether opening a new tab should open the Home page
+// instead of the New Tab page.
+const char kNewTabPageIsHomePage[] = "newtabpage_is_homepage";
+
 // This is the URL of the page to load when opening new tabs.
 const char kHomePage[] = "homepage";
 
--- a/chrome/common/pref_names.h
+++ b/chrome/common/pref_names.h
@@ -323,6 +323,7 @@ extern const char kSettingsShowOSBanner[
 extern const char kDeviceLoginScreenWebUsbAllowDevicesForUrls[];
 #endif  // defined(OS_CHROMEOS)
 extern const char kShowHomeButton[];
+extern const char kNewTabPageIsHomePage[];
 extern const char kSpeechRecognitionFilterProfanities[];
 extern const char kAllowDeletingBrowserHistory[];
 #if !defined(OS_ANDROID)
